backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -428,6 +433,120 @@ public FutureExperimentRunDAO(
+              for (var entry : datasetVersionDatasetList) {
+                datasetVersionDatasetMap.put(entry.getKey(), String.valueOf(entry.getValue()));
+              }
+              return InternalFuture.completedInternalFuture(datasetVersionDatasetMap);
+            },
+            executor)
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              // Validate linked dataset in dataset version exists in database records if not then
+              // we will throw the error
+              Set<String> datasetIds = new HashSet<>();
+              for (Artifact dataset : datasets) {
+                String datasetVersionId = dataset.getLinkedArtifactId();
+                if (datasetVersionDatasetMap.containsKey(datasetVersionId)
+                    && datasetVersionDatasetMap.get(datasetVersionId) != null) {
+                  datasetIds.add(datasetVersionDatasetMap.get(datasetVersionId));
+                } else {
+                  return InternalFuture.failedStage(
+                      new InvalidArgumentException(
+                          "Dataset not found for dataset version: " + datasetVersionId));
+                }
+              }
+
+              // If all dataset with linked dataset versions found in database then checking
+              // authorization for the current user for those datasets
+              List<InternalFuture<String>> internalFutures = new LinkedList<>();
+              for (String datasetId : datasetIds) {
+                try {
+                  internalFutures.add(
+                      InternalFuture.completedInternalFuture(datasetId)
+                          .thenCompose(
+                              id ->
+                                  checkEntityPermissionBasedOnResourceTypes(
+                                          Collections.singletonList(datasetId),
+                                          ModelDBActionEnum.ModelDBServiceActions.READ,
+                                          ModelDBServiceResourceTypes.DATASET)
+                                      .thenCompose(
+                                          unused -> InternalFuture.completedInternalFuture(id),
+                                          executor),
+                              executor));
+
+                } catch (PermissionDeniedException ex) {
+                  LOGGER.trace(ex.getMessage());
+                  // We are throws error while creating entities
+                  // if we are fetching entities then we are just skip those datasets from response
+                  // and skipped PermissionDeniedException on fetch
+                  if (errorOut) {
+                    return InternalFuture.failedStage(ex);
+                  }
+                }
+              }
+
+              // If all mapped datasets are allowed for the user then we will filter datasets based
+              // on linked dataset version and allowed dataset ids.
```

Comment:
Why do we have to filter if they are all allowed? This seems incorrect to me

Reply:
they are all allowed means the all allowed dataset ids but using those ids i can not directly filter the datasets of runs so i have to applied map logic for linked_artifact_id and datasetId and then based on allowed ids i have to get those Linked datasetVersionId and want to filter with ER dataset `linked_artifact_id`